000X-R-In-tree-pluggable-components.md

```diff
@@ -0,0 +1,132 @@
+
+## Related Items
+
+### Related proposals
+
+This proposal is related to the "pluggable component" techology that was first implemented with Dapr 1.9, which is still evolving.
+
+We've considered the issues of the large number of components in the past, with proposals such as "cloud-specific builds", for example dapr/dapr#3168. However, those proposals came with a lot of limitations, both for Dapr developers and end-users.
+
+## Expectations and alternatives
+
+* **What is in scope for this proposal?**
+  * This proposal is about the underlying technology to support instantiating in-tree components as pluggable.
+* **What is deliberately *not* in scope?**
+  * This is *not* a proposal about removing components from Dapr so they are not shipped together with Dapr and in the same Docker containers.
+  * This proposal is also *not* concerned with *which* components should be shipped as statically compiled or in-tree pluggable. Maintainers are expected to make this decision collegially in separate conversations.
+  * This is also *not* a proposal for a technology that enables customers to build images of Dapr with a custom set of components statically compiled or as in-tree pluggable, although this does enable that scenario in the future.
+* **What alternatives have been considered, and why do they not solve the problem?**
+  * The first alternative is the *status quo*. Dapr continues to grow with the addition of new components.
+  * We can also decide to ship multiple "flavors" of Dapr each with a different set of components statically compiled inside, e.g. as per dapr/dapr#3168. This however is more limiting for end users and adds operational complexities for Dapr developers.
+  * Removing components from Dapr entirely is an option as well, for example to ship them as external, pluggable components. These would be separate from `daprd` in all ways, including: they're shipped in separate Docker images, need to be explicitly added to applications (see [0005-R-pluggable-components-injector.md](./0005-R-pluggable-components-injector.md)), and can have separate release lifecycles and support.
+  * Users can also re-compile Dapr with a custom set of components. Although we don't officially support this, we are aware of at least some users who do this as a practice. This is something only advanced users may consider doing due to the very high complexity and operational burden.
+* **Are there any trade-offs being made?**  
+  * Components that are instantiated as in-tree pluggable components will likely have *slightly* degraded performance, due to the overhead of inter-process communication and gRPC serialization/unserialization. Based on the research that was done for pluggable components, that should be negligible for most users.
```

Comment:
It woud be great if we could quantify this but I don't think we ever came with those numbers in  pluggable components final incarnation.